### PR TITLE
Create patients

### LINKED ISSUE #601: Allow the user to register a new patient
Include:
- Name
- Birthday
- Location
- Address
- Email
- Phone

Country and State are not PII.

All fields should be stored within the core fields.

Location should behave like site location.

### LINKED ISSUE #608: [hiereorgs] Scope patients crud

### PR DESCRIPTION
**Pull request by bcardiff**
Note: Created to start an early review cc: @nekron @spalladino 

* fixes #601 adds patients related permissions
* fixes #608 using only the institution of the navigation context
* DRY validation_errors
* DRY pagination controls
* Use current user locale in requests
* basic patient listing, just paging.

### CODE REVIEW

File: app/controllers/patients_controller.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+class PatientsController < ApplicationController
+  def index
+    @can_create = has_access?(@navigation_context.institution, CREATE_INSTITUTION_PATIENT)
+    @patients = check_access(Patient.where(institution: @navigation_context.institution), READ_PATIENT).order(updated_at: :desc)
+  end
```
Comment (spalladino) [inline app/controllers/patients_controller.rb]: If patients don't have a site, then the only way for a user to browse patients is to have permissions over the whole institution. Should we add something more related to site?

File: app/models/patient.rb [hunk 2]
```diff
@@ -22,4 +23,37 @@ def has_entity_id?
...
+          else
+            self.core_fields[arg.to_s] = value
+          end
+        end
+      end
+    end
+  end
+
+  attribute_field :name, :gender, :dob, :email, :phone
+
```
Comment (spalladino) [inline app/models/patient.rb]: Looks nice. Should we move this to Entity?

File: app/models/patient_form.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    save
+  end
+
+  def save
+    self.class.assign_attributes(patient, self)
+    self.valid? && patient.save
+  end
+
+  validates_presence_of :name
+  GENDER_VALUES = %w(male female other)
```
Comment (spalladino) [inline app/models/patient_form.rb]: `Cdx::Fields.entities.core_field_scopes.find{|s| s.name == 'patient'}.fields.find{|f| f.name == 'gender'}.options`

File: app/models/patient_form.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  def dob_text_placeholder
+    date_format[:placeholder]
+  end
+
+  def dob_is_a_date
+    return if dob.blank?
+    errors.add(:dob_text, "should be a date in #{dob_text_placeholder}") unless dob.is_a?(Time)
+  end
+  # end dob
```
Comment (spalladino) [inline app/models/patient_form.rb]: What about keeping a single field "dob" which is typecasted to datetime when read, like:
```ruby
def available_at
  return super.to_datetime if super.respond_to?(:to_datetime)
  DateTime.parse(super)
end
```

Then validating the format client-side, with a `pattern` attribute, and just the presence here?

File: app/views/device_models/_form.haml [hunk 5]
```diff
@@ -10,13 +10,7 @@
...
   });
 
-- if @device_model.errors.present?
-  .alert-container
-    .alert.alert-error
-      %ul
-        - if @device_model.errors.present?
-          - @device_model.errors.full_messages.each do |msg|
-            %li= msg
+= validation_errors @device_model
```
Comment (spalladino) [inline app/views/device_models/_form.haml]: :smiley:

File: app/models/patient.rb (hunk 2 above)
Comment (bcardiff) [inline app/models/patient.rb]: We can, yet is only used in Patient at the moment. Also we need to add a way to copy a field to an AR attribute.  Maybe a `copy: true` as an options. That will pollute a bit the above code most likely.

File: app/models/patient_form.rb (hunk 3 above)
Comment (bcardiff) [inline app/models/patient_form.rb]: `Patient.entity_fields.detect { |f| f.name == 'gender' }.options`, but yeah +1

File: app/models/patient_form.rb (hunk 4 above)
Comment (bcardiff) [inline app/models/patient_form.rb]: :-1: regarding "just the presence". The client-side pattern is not enough.

But I could try to keep it just in a single property. I make the two due to previous intent of wrapping the formatting in patient model directly with alias_method_chain and the getter/setter property just got to ugly.

Related technical debt: dob is stored/read as text from Entity.

**Pull request merged**